products/libllbuild/Core-C-API.cpp

```diff
@@ -240,6 +273,36 @@ void llb_buildengine_task_is_complete(llb_task_interface_t* ti_p,
   ti->complete(std::move(result));
 }
 
+// FIXME: We need to expose customizing the JobDescriptor somehow. Can't be
+// nullptr since those are used as sentinels that the execution queues are
+// finishing.
+class CAPIJobDescriptor : public JobDescriptor {
+public:
+  CAPIJobDescriptor() {}
+
+  virtual StringRef getOrdinalName() const {
+      return StringRef("");
+  }
+
+  virtual void getShortDescription(SmallVectorImpl<char> &result) const {
+      llvm::raw_svector_ostream(result) << getOrdinalName();
+  }
+
+  virtual void getVerboseDescription(SmallVectorImpl<char> &result) const {
+      llvm::raw_svector_ostream(result) << getOrdinalName();
+  }
+};
+
+void llb_buildengine_spawn(llb_task_interface_t* ti_p, void* context, void (*action)(void* context, llb_task_interface_t* ti_p)) {
+    auto ti = ((TaskInterface*) ti_p);
+    auto ti_cp = TaskInterface(*ti);
```

Comment:
this line right here...  it will haunt me forever, I hate C++

Reply:
Ah but the grass on the other side could be something like:

```
ti_p.withMemoryRebound(to: UnsafePointer<TaskInterface>.self, capacity: 1) { ti in
  let ti_cp = TaskInterface(copying: ti)
  ...
}
```